create backups via UI

machisuji opened the pull request:
WP [#35645](https://community.openproject.org/projects/saas/work_packages/35645/activity)

machisuji committed:
create backups via UI

oliverguenther committed:
Fix import of modal service

oliverguenther commented on Gemfile:
```diff
@@ -37,6 +37,10 @@ gem 'activerecord-session_store', '~> 2.0.0'
 
+# Keep mimemagic at older version until rails provides its own
+# solution without adding a system dependency. We are GPL ourselves so this is fine.
+gem 'mimemagic', git: 'https://github.com/opf/mimemagic', ref: 'bf8d7c2'
```
Can be removed in dev now

oliverguenther commented on app/models/export.rb:
```diff
@@ -0,0 +1,12 @@
+class Export < ApplicationRecord
+  has_one(
```
:+1: for extracting this

oliverguenther commented on app/services/concerns/contracted.rb:
```diff
@@ -36,8 +36,8 @@ module Contracted
     def contract_class=(cls)
-      unless cls <= ::BaseContract
-        raise ArgumentError "#{cls.name} is not an instance of BaseContract."
+      unless cls <= ::ModelContract
```
Why did you change this back? I'm pretty sure I originally moved this from ModelContract to BaseContract in case we have a service that doesn't depend on an AR model

machisuji replied:
All I did was add a missing comma there after the `ArgumentError` as far as I'm concerned.

machisuji replied:
But yeah apparently the class changed too there. Since I based this on release/11.2 originally this might've been mixed up. I'll change the class back.

oliverguenther commented on app/views/admin/backups/show.html.erb:
```diff
@@ -0,0 +1,39 @@
+
+<%= toolbar title: t(:label_backup) %>
+
+<backup
```
Just a preference but I'd prefer a

```
tag :backup, data: { 'job-status-id': .... }
```

as that handles escaping the attributes correctly in case you ever want to pass a string with a quote in it

oliverguenther commented on frontend/src/app/components/api/op-backup/op-backup.service.ts:
```diff
@@ -0,0 +1,55 @@
+import {filter, map, share} from "rxjs/operators";
+import {HalResourceService} from "core-app/modules/hal/services/hal-resource.service";
+
+@Injectable()
```
```suggestion
@Injectable({ providedIn: 'root' })
```

Instead of providing in the uppermost module angular4 modules, please use this instead. This will result in the service becoming a global singleton whenever any component or service requires it.

oliverguenther commented on lib/api/v3/backups/backups_api.rb:
```diff
@@ -0,0 +1,82 @@
+              fail ::API::Errors::Conflict, message: "There is already a backup pending."
+            end
+
+            limit = OpenProject::Configuration.backup_daily_limit
```
I'd probably want to have this moved into the contract level. This ensures it's being validated everywhere and you can react to it with a special response in case of this error symbol.

oliverguenther commented on spec/features/admin/backup_spec.rb:
```diff
@@ -0,0 +1,62 @@
+
+    expect(page).to have_text "The export has completed successfully"
+    expect(subject).to end_with ".zip"
+  end
```
Missing a spec to test the 429 when the daily limit is being hit

machisuji replied:
I've added a spec testing a general (400) error to the feature specs. The specific errros (429, 403 etc.) are covered in the backup API spec.

oliverguenther reviewed (changes requested):
I got some minor comments.

Additionally, these followings things I noticed:

- [x] The export does not include the attachments even though I requested them through the checkbox
- [x] The export can be triggered by anyone with the permission and doesn't contain the discussed "first-time generated token" thing?

machisuji committed:
introduced backup token and addressed remaining comments

machisuji committed:
allow disabling permissions

machisuji commented:
@oliverguenther Could you please have another look? This now does include the backup token. Though I've changed it so it's not just generated once initially but can be regenerated. Open for discussion I guess but I think the way it is now will give us less trouble (work) later on because people will be able to reset the tokens themselves.

To avoid the idle desktop scenario we talked about a newly generated token cannot be used straight away and also will trigger an email informing the actual user that a new token has been generated. If it wasn't them this gives them the chance to react and reset it again.

oliverguenther commented:
Functionality-wise, LGTM now. As a user of the feature I'm still a bit confused by the delay as I do not get informed about that when first creating the backup token. Only when I start using it will I be greeted with an error that I still have to activate it.

I'd appreciate it if we extend the info on the page a bit, and also add a piece of documentation about how to trigger the allow_now.

I had a local installation with more than 1GB of attachments, which is why the attachments were not added. This is also not clear to the user and the attachments are just missing. I'd also appreciate either disabling the checkbox or some other measure when the attachment file size exceeds the possible backup size.

oliverguenther commented on app/controllers/admin/backups_controller.rb:
```diff
@@ -0,0 +1,92 @@
+  end
+
+  def reset_token
+    if request.post?
```
I generally don't like the `if request.post?` too much. How about two separate routes for get/post with two different actions?

oliverguenther commented on frontend/src/app/components/admin/backup.component.ts:
```diff
@@ -0,0 +1,115 @@
+@Component({
+  selector: backupSelector,
+  templateUrl: './backup.component.html',
+  styleUrls: ['./backup.component.sass']
```
You don't have to create a .sass entry here if you don't have any styles. You can just remove it.

oliverguenther reviewed (changes requested):
Only some minor comments code-wise

machisuji committed:
improvements

- only make user wait to use backup token in if really necessary
- notify admins of new backup token
- disable 'include attachments' option in UI if unavailable
- documentation
- misc

machisuji committed:
spec fixes

machisuji committed:
fixed feature spec

machisuji committed:
allow setting capybara host in every case

machisuji committed:
removed unused style file

machisuji committed:
addressed review feedback, added further feature specs

machisuji committed:
polish (code climate)

oliverguenther committed:
Avoid empty attachments

oliverguenther committed:
Don't raise filesize validation for internal exports

oliverguenther reviewed (approved):
LGTM :+1: I observed two corner cases where attachments were missing which shouldn't usually happen but I still pushed a miner fix for that.

Also, on my instance the export would exceed the attachment file size, so I added a skip check for the filesize in case the attachment is an export. Since you based Backups on the Export class, this will also fix it for work package or other exports.
